backend/sqlite/sqlite.go

```diff
@@ -390,20 +392,55 @@ func (be *sqliteBackend) CompleteOrchestrationWorkItem(ctx context.Context, wi *
 }
 
 // CreateOrchestrationInstance implements backend.Backend
-func (be *sqliteBackend) CreateOrchestrationInstance(ctx context.Context, e *backend.HistoryEvent) error {
+func (be *sqliteBackend) CreateOrchestrationInstance(ctx context.Context, e *backend.HistoryEvent, option *protos.CreateInstanceOption) error {
 	if err := be.ensureDB(); err != nil {
 		return err
 	}
 
+	// if user not set this option, it will be nil, initialize it here to set default value
+	if option == nil {
+		option = &protos.CreateInstanceOption{}
+	}
+
 	tx, err := be.db.BeginTx(ctx, nil)
 	if err != nil {
 		return fmt.Errorf("failed to start transaction: %w", err)
 	}
 	defer tx.Rollback()
 
 	var instanceID string
-	if err := be.createOrchestrationInstanceInternal(ctx, e, tx, &instanceID); err != nil {
-		return err
+	runtimeStatus, instanceID, err := be.createOrchestrationInstanceInternal(ctx, e, tx)
+	if err != nil {
+		// instance alredy exists
+		if errors.Is(err, backend.ErrDuplicateEvent) {
+			// build target status set
+			statusSet := buildStatusSet(option.OperationStatus)
+			// if current status is not one of the target status, return error
+			if _, ok := statusSet[helpers.FromRuntimeStatusString(runtimeStatus)]; !ok {
+				return api.ErrDuplicateInstance
+			} else {
+				switch option.Action {
+				case protos.CreateOrchestrationAction_SKIP:
+					// Log an warning message and skip cerating new instance
+					be.logger.Warnf("An instance with ID '%s' already exists; dropping duplicate create request", instanceID)
+					return nil
+				case protos.CreateOrchestrationAction_TERMINATE:
+					// terminate existing instance
+					if err := be.cleanupOrchestrationStateInternal(ctx, tx, api.InstanceID(instanceID)); err != nil {
+						return err
+					}
+					// create a new instance
+					if _, _, err := be.createOrchestrationInstanceInternal(ctx, e, tx); err != nil {
+						return err
+					}
+				case protos.CreateOrchestrationAction_ERROR:
+					// CreateInstanceAction_ERROR
+					return api.ErrDuplicateInstance
+				}
+			}
+		} else {
+			return err
+		}
 	}
```

Comment:
Flattening it out a bit to improve code readability. Added a comment "Do not return here" to indicate the omission of a return statement was intentional:


```suggestion
	// Instance alredy exists
	if errors.Is(err, backend.ErrDuplicateEvent) {
		// build target status set
		statusSet := buildStatusSet(option.OperationStatus)

		// if current status is not one of the target status, return ErrDuplicateInstance
		if _, ok := statusSet[helpers.FromRuntimeStatusString(runtimeStatus)]; !ok {
			return api.ErrDuplicateInstance
		}

		switch option.Action {
		case protos.CreateOrchestrationAction_SKIP:
			// Log a warning message and skip creating a new instance
			be.logger.Warnf("An instance with ID '%s' already exists; dropping duplicate create request", instanceID)
			return nil
		case protos.CreateOrchestrationAction_TERMINATE:
			// Terminate the existing instance
			if err := be.cleanupOrchestrationStateInternal(ctx, tx, api.InstanceID(instanceID)); err != nil {
				return err
			}

			// Create a new instance
			if _, _, err := be.createOrchestrationInstanceInternal(ctx, e, tx); err != nil {
				return err
			}

			// Do not return here - let this fall through
		case protos.CreateOrchestrationAction_ERROR:
			// CreateInstanceAction_ERROR
			return api.ErrDuplicateInstance
		}
	} else if err != nil {
		return err
	}
```

Reply:
updated